chat-app: cleanup log noise, address deprecation messages

mamuqsit opened the pull request:
## Introduction
Excessive log output and deprecated messages here create noise during development and trigger warnings in recent dependency versions. Log spam reduce readability and raise deprecation warnings in console / Log stream.

## Changes
- Suppress noisy “Transmission succeeded” and similar debug logs c85ce30e521f7bed1c57ed070077b48b4dab801a
- Add clean_message_for_console() to remove binary audio data from console logs 1148a38beb3d8772d1dbc51cec7d958a0450742a 27e70d4d205591cbb4e3547d968ff707824a7838
- Modernize Pydantic model definitions using examples= and ConfigDict 9922d412dae2de63bb1f6f0e550aa68575e48406 56bbe7d69b8859a9d81f1954d8786ad9dcf779c5
- Use the websockets-sansio implementation in Uvicorn fe00c94a62b5da57d81533b101c070bb327565c1
- ~~Update openai, llama-index-* dependency versions in byoeb-core and byoeb-integrations be7efdf04a8a0d425d4f856b1d15f5778e268609~~
- Fix bad cache logic in [workflows/tests.yml](https://github.com/A4i-tech/byoeb/pull/59/files#diff-1db27d93186e46d3b441ece35801b244db8ee144ff1405ca27a163bfe878957f) - a redundant 'Install dependencies' step was run prior to cache (meaning we didn't properly used cache for byoeb-v1/byoeb)
- Remove unused deps ("notebook") 710a636910bb10807e3e8d064c8d245384379fa0

mamuqsit committed:
chat/utils: implement clean_message_for_console - remove raw audio bytes from messages

mamuqsit committed:
user-flow-handler: print messages without audio content

mamuqsit committed:
chat-app: suppress 'Transmission succeeded' logs

mamuqsit committed:
pydantic: example= -> examples=
"examples" is the official parameter name per https://docs.pydantic.dev/latest/api/fields/

This addresses several of these following deprecation message that appear on startup:
```
  recipient_id: Optional[str] = Field(None, description="Unique identifier for the recipient", example="user123")
byoeb_core/models/byoeb/message_status.py:12: PydanticDeprecatedSince20: Using extra keyword arguments on `Field` is deprecated and will be removed. Use `json_schema_extra` instead. (Extra keys: 'example'). Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.12/migration/
```

mamuqsit committed:
pydantic: replace nested Config class with ConfigDict

Source: https://docs.pydantic.dev/2.12/migration/
> In Pydantic V2, to specify config on a model, you should set a class attribute called model_config to be a dict with the key/value pairs you want to be used as the config.
> The Pydantic V1 behavior to create a class called Config in the namespace of the parent BaseModel subclass is now deprecated.

Source: https://docs.pydantic.dev/2.12/api/config/#pydantic.config.ConfigDict.populate_by_name
> populate_by_name usage is not recommended in v2.11+ and will be deprecated in v3. Instead, you should use the validate_by_name configuration setting.
> When validate_by_name=True and validate_by_alias=True, this is strictly equivalent to the previous behavior of populate_by_name=True.

This addresses these deprecation messages on startup:
```
byoeb_core/models/whatsapp/incoming/interactive_message.py:52: PydanticDeprecatedSince20: Support for class-based `config` is deprecated, use ConfigDict instead. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.12/migration/
  class WhatsAppInteractiveMessageBody(BaseModel):
```

mamuqsit committed:
fastmcp: provide stateless_http in http_app() instead
This addresses the following deprecation message on startup:
```
fastmcp/server/server.py:255: DeprecationWarning: Providing `stateless_http` when creating a server is deprecated. Provide it when calling `run` or as a global setting instead.
```

mamuqsit committed:
uvicorn: use websockets-sansio impl
Although ASHAbot does not utilize websockets, uvicorn loads the deprecated ws module on startup.

Source: https://uvicorn.dev/concepts/websockets/#websockets-sansio-protocol
> Since websockets deprecated the API Uvicorn uses to run the previous protocol, we had to create this new protocol that uses the websockets SansIO API.
> You can choose this protocol by setting the --ws option to websockets-sansio.

This addresses the following deprecation message on startup:
```
websockets/legacy/__init__.py:6: DeprecationWarning: websockets.legacy is deprecated; see https://websockets.readthedocs.io/en/stable/howto/upgrade.html for upgrade instructions
  warnings.warn(  # deprecated in 14.0 - 2024-11-09
uvicorn/protocols/websockets/websockets_impl.py:17: DeprecationWarning: websockets.server.WebSocketServerProtocol is deprecated
  from websockets.server import WebSocketServerProtocol
```

mamuqsit committed:
poetry: update openai, llama-index-* deps

Pydantic deprecation messages originate from us using outdated library versions.

This addresses the following deprecation message on startup:
```
pydantic.warnings.UnsupportedFieldAttributeWarning: The 'validate_default' attribute with value True was provided to the `Field()` function, which has no effect in the context it was used. 'validate_default' is field-specific metadata, and can only be attached to a model field using `Annotated` metadata or by assignment. This may have happened because an `Annotated` type alias using the `type` statement was used, or if the `Field()` function was attached to a single member of a union type.
```

mamuqsit committed:
pyproject: temporarily use 'reduce-log-spam' for workflow

mamuqsit committed:
I wonder whether this is a cache issue... ?

mamuqsit committed:
test-vector-store-chroma: use llama-provided mocks for embeddings

mamuqsit committed:
tests: specify dummy AZURE_OPENAI_API_KEY so we can import dep setup successfully

mamuqsit committed:
workflow: fix improper use of cache

mamuqsit committed:
poetry: remove unused dep "notebook"

Soumabha commented on byoeb-v1/byoeb-integrations/byoeb_integrations/vector_stores/tests/test_vector_store_chroma.py:
```diff
@@ -1,104 +1,34 @@
-from byoeb_integrations.embeddings.llama_index.azure_openai import AzureOpenAIEmbed
+from chromadb import Documents, EmbeddingFunction, Embeddings
 from byoeb_integrations.vector_stores.chroma.base import ChromaDBVectorStore
-from byoeb_integrations.embeddings.chroma.llama_index_azure_openai import AzureOpenAIEmbeddingFunction
```
Since we might get merge conflicts for the changes to the vector store (refer: https://github.com/A4i-tech/byoeb/pull/60), could we do it directly as a PR atop that branch instead and remove it from the current changes for the cleanup.

I agree with the usage of the MockEmbeddings instead of putting the AzureOpenAIEmbedding (as is happening incorrectly in the other branch)

mamuqsit replied:
Moved to https://github.com/A4i-tech/byoeb/pull/62 and reverted changes here https://github.com/A4i-tech/byoeb/pull/59/commits/3c5c9ca332bb49ad00cf574020cae3372e3ff989

Soumabha commented on byoeb-v1/byoeb-integrations/pyproject.toml:
```diff
@@ -8,17 +8,17 @@ exclude = ["**/tests/**", "**/tests"]
+llama-index-vector-stores-chroma = "0.5.3"
 azure-search-documents = "11.5.2"
-byoeb-core = { git = "https://github.com/A4i-tech/byoeb.git", rev = "a4i/main", subdirectory = "byoeb-v1/byoeb-core" }
+byoeb-core = { git = "https://github.com/A4i-tech/byoeb.git", rev = "reduce-log-spam", subdirectory = "byoeb-v1/byoeb-core" }
```
Im considering just leveraging local reference itself. Do you see any challenges with that?

mamuqsit replied:
Should be quite straightforward - no challenges.

Docker cannot copy files from `..` (parent directory), so we would move Dockerfile 1 folder up (in `byoeb-v1` folder), and then set `WORKDIR /app/byoeb`

Soumabha replied:
Lets take this up separately then as we will need to come up with a solution and them implement it. Approved the changes, please correct it before merge.

Soumabha commented on byoeb-v1/byoeb/pyproject.toml:
```diff
@@ -9,15 +9,14 @@ readme = "README.md"
 fastapi = "^0.115.5"
 aiocache = "0.12.3"
-byoeb-integrations = {git = "https://github.com/A4i-tech/byoeb.git", branch = "a4i/main", subdirectory = "byoeb-v1/byoeb-integrations"}
+byoeb-integrations = {git = "https://github.com/A4i-tech/byoeb.git", branch = "reduce-log-spam", subdirectory = "byoeb-v1/byoeb-integrations"}
```
Im planning to move to use local references via ../ itself. Do you see any issues with that?

Soumabha replied:
Lets take this up separately then as we will need to come up with a solution and them implement it. Approved the changes, please correct it before merge.

Soumabha commented on byoeb-v1/byoeb/tests/services/test_chat_utils.py:
```diff
@@ -0,0 +1,41 @@
+from byoeb.services.chat.utils import clean_message_for_console
```
Name should be test-chat-utils

mamuqsit replied:
Addressed in https://github.com/A4i-tech/byoeb/pull/59/commits/8e3e532240eaf827b60c38a02d4529ee408e4cc8

mamuqsit committed:
rename: test_chat -> test_chat_utils

mamuqsit committed:
Revert "test-vector-store-chroma: use llama-provided mocks for embeddings"

This reverts commit 57b88d8fb45d7795207a28ca33c9cc690e7ca203.

a4i-architect reviewed (commented):
PR Review Summary:

Overall, the code changes reflect incremental improvements primarily focused on:

- Standardizing example and example list annotations in Pydantic models.
- Minor adjustments to imports and configuration handling.
- Updating dependency versions.
- Enhancing logging configuration.
- Refactoring the `clean_message_for_console` utility for more robust message sanitization.
- Updating test cases to cover new behaviors.

Positives:
- The use of `examples` as a list in Pydantic enhances API documentation clarity.
- The `clean_message_for_console` function now thoroughly removes binary data, which is a good practice for console logging and debugging.
- Version upgrades (e.g., `poetry`, dependencies, Uvicorn) bring improvements in performance, security, and features.
- Logging level adjustments help suppress verbose logs, making logs cleaner.

Identified issues and suggestions:
1. **Redundant Configuration in Pydantic Models (models/whatsapp/incoming/interactive_message.py & regular_message.py):**  
   The `model_config = ConfigDict(...)` attribute within the class body is a modern, correct approach, but its mix with a commented-out `Config` class shows inconsistency. Prefer sticking to one approach.  
   *Recommendation:* Remove the outdated `Config` class entirely for clarity.  

2. **Typing and Annotations for List Examples:**  
   The usage of `examples=["value"]` for fields like `message_id`, `message_category`, etc., is correct for documentation, but to be fully expressive, consider specifying `example=["val1", "val2"]` only when multiple examples are relevant. For single examples, `example="value"` suffices, or `examples=["val"]` if multiple are intended. This is minor but worth standardizing.

3. **Duplicate/Unnecessary Import and Code:**  
   The changes to import statements (`config`) and minor logging adjustments are fine, but avoid leaving commented or inconsistent remnants (e.g., the `Config` class instead of `model_config`).

4. **Version Bumps & Dependency Management:**  
   Upgrades to dependencies like `llama-index`, `openai`, and `uvicorn` are well justified, especially to benefit from newer features and security patches.  
   *Suggestion:* Keep dependencies pinned precisely, possibly review major version changes for compatibility.

5. **Code Complexity & Testing:**  
   The `clean_message_for_console()` utility is well-crafted. The test case is comprehensive, checking for proper removal of binary data and object integrity.  
   *Note:* The function currently copies the message object deeply and manipulates `additional_info`. This approach seems sufficient; no over-complication is evident.

6. **Logging & Initialization Logic**  
   The modifications to the app startup (main.py `run.py`) for logging and WebSocket settings are straightforward. Increasing `ws` parameter clarity improves WebSocket support.

7. **Unrelated/Minor Style:**  
   The mix of `description` vs `examples` annotations suggests a need for standardization. Prefer `examples` list when multiple examples are relevant, else keep to a string.

**Web Search Articles & Best Practices:**
- Using Pydantic’s `examples` attribute improves documentation quality: [Pydantic Docs - Field](https://pydantic.dev/usage/schema/#example).
- Cleaning binary data for logging aligns with security best practices: [Logging Binary Data](https://docs.python.org/3/library/logging.html#logging.LogRecord).
- Dependency management and version pinning are critical for stability, especially in production: [Dependency Versioning Best Practices](https://realpython.com/dependency-management-python/).

**Final Score: 75/100**

**Summary:**
The modifications are generally sound and pragmatic, aligning with best practices in logging, API documentation, and dependency management. Minor improvements around code consistency, documentation standards, and removal of redundant code could elevate the quality further. The core architecture remains simple, with no overly complicated logic or high cyclomatic complexity apparent.

**Acceptance Note:** The current state meets basic standards with room for polish.

---

mamuqsit committed:
tests: add dummy AZURE_OPENAI_API_KEY to silence errors

If https://github.com/A4i-tech/byoeb/pull/62 is merged before this PR, accept incoming changes.
If this PR is merged first, overwrite this PR's changes.

a4i-architect reviewed (commented):
PR Review Summary:

Overall, the code changes reflect a mix of structural improvements, configuration updates, and minor code cleanup, with an emphasis on better API modeling, environment setup, and test robustness. Below is a detailed critique and suggestions for improvement:

1. **Workflow Adjustment (tests.yml)**:
   - Simplification of the Python setup step by removing the explicit dependency installation (`poetry install`) is appropriate if the dependencies are cached or pre-installed. However, ensure this is compatible with your CI/CD pipeline.
   - Consider explicitly testing cache effectiveness by checking cache hits/misses or documenting why the step was removed, to avoid potential build inconsistencies.

2. **Pydantic Models (message_context.py)**:
   - Transitioning to `examples` instead of `example` is reasoned, yet the standard from Pydantic is `example`. Using `examples` might lead to compatibility issues or confusions with documentation tools. Verify if Pydantic 1.x or 2.x supports this; typically, `example` is preferred.
   - Numerous string fields switch from `example` to `examples=[...]`. This change is inconsistent; stick to `example` for singular examples for simplicity unless multiple are needed.
   - The removal of newline at EOF is good practice.

3. **Message Status Models**:
   - The update to `examples` mirrors the above comments. Revert to `example` unless `examples` is explicitly required by your validation or documentation tools.

4. **User Model**:
   - The `user_location` field changes to a dict with default `{}`. Remember, mutable defaults are discouraged; instead, use `default_factory=dict` for safety to avoid shared mutable defaults. This applies across model classes to prevent potential bugs.

5. **Incoming/Regular Message Models**:
   - The introduction of `model_config = ConfigDict(validate_by_name=True, validate_by_alias=True)` is a notable change. Confirm version compatibility; typically, Pydantic’s `Config` class is used. This might be a new API or custom wrapper—document it for maintainability.
   - The switch from nested `Config` class to `model_config` is potentially a simplification, but it’s crucial that the codebase consistently supports this API style.

6. **Poetry Dependency Updates**:
   - Updating dependencies (e.g., llama-index, openai) to newer versions is beneficial for bug fixes and features. Make sure to test thoroughly since minor version upgrades, especially from 0.11.23 to 0.14.7, could introduce breaking changes.

7. **Azure Environment Variable Setup for Tests**:
   - Setting `AZURE_OPENAI_API_KEY` directly in test files is pragmatic but should be documented as a temporary workaround. For security, avoid hardcoding secrets in code or tests.
   - Consider using environment variable configuration management or secret vault integrations.

8. **Main Application (run.py)**:
   - The removal of `stateless_http=True` from `FastMCP()` constructor might change middleware behavior; verify if this impacts sessionless communication.
   - The logging configuration change to suppress verbose Azure logs (`azure.monitor.opentelemetry.exporter.export._base`) improves log clarity.
   - The WebSocket `ws="websockets-sansio"` parameter consistency is good but document its necessity.
   - Overall, the server runtime configuration is more streamlined, but ensure these changes are tested for impact on production deployment stability.

9. **Message Handling (generate.py)**:
   - The addition of `clean_message_for_console()` demonstrates better logging hygiene, especially when dealing with binary data.
   - Use of `model_copy(deep=True)` is safe but might add slight overhead; evaluate if a shallow copy suffices.
   - The `if byoeb_expert_message else byoeb_expert_message` check is redundant; just passing the variable should be fine unless `None` needs specific handling.

10. **Utility Function (utils.py)**:
    - The `clean_message_for_console()` function efficiently removes binary data from `additional_info`. Confirm the intended use case: if other nested structures contain binary data, consider extending logic.
    - Defensive programming—ensure the structure of `additional_info` is always as expected to avoid runtime errors.

11. **General Observations**:
    - Consistency in use of `example` vs `examples`.
    - Avoid mutable default arguments (`{}` or `[]`) in Pydantic models—prefer `default_factory`.
    - Optionally add type hints for complex structures or enforce explicit validation schemas.
    - The PR improves logging, dependency management, and model clarity but can benefit from a more uniform approach to Pydantic configuration.

**Web Articles & References**:
- Pydantic `example` vs `examples`: https://pydantic.dev/latest/usage/models/#field-customization
- Python mutable default arguments: https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments
- Effective logging practices in Python: https://realpython.com/python-logging/
- Managing secrets in tests: https://docs.microsoft.com/en-us/azure/devops/pipelines/library/variable-groups?view=azure-devops

**Score: 78/100**  
*The changes are positive and improve configuration, model clarity, and test robustness, but inconsistencies in Pydantic field annotations, minor structural redundancies, and environment handling slightly reduce the overall score.*  
Further minor refactoring for uniformity and safeguarding defaults would push it toward the acceptable threshold for production readiness.

**Recommendations**:
- Revert `examples` to `example` unless your project explicitly benefits from `examples`.
- Use `default_factory=dict` or `default_factory=list` for mutable defaults.
- Ensure all class configs follow a consistent Pydantic pattern (`Config` class or `model_config`).
- Document environment setup for tests for better maintainability.

This approach enhances code quality, consistency, and clarity for future development and reviews.

mamuqsit committed:
revert dependency ugrades

a4i-architect reviewed (commented):
### PR Review Summary

#### General Overview
The proposed code modifications primarily focus on enhancing type annotations, cleaning up configuration, and implementing a utility function for message sanitization. Most changes aim to improve data clarity, consistency, and logging control. Overall, the refactoring is sensible but contains some areas for improvement.

---

### Detailed Feedback

#### 1. **Workflow YAML (`tests.yml`)**
- The removal of `poetry install` and move to a more minimal setup could be acceptable if dependencies are pre-cached or managed elsewhere. However, **removing dependency installation** might lead to flaky tests if the cache misses.
- The explicit setting of `stateless_http=True` in `FastMCP()` was perhaps redundant; setting it during instantiation clarifies intent. The change to `mcp = FastMCP()` without arguments could be an oversight unless defaults are well understood.
  
**Suggestion:** Verify that dependency installations happen reliably outside this workflow step if removing `poetry install`. Clarify the defaults used for `FastMCP()`.

#### 2. **Model Files (`message_context.py`, `message_status.py`, `user.py`, etc.)**
- The consistent use of `examples` instead of `example` in `Field()` is correct based on Pydantic v2 since `examples` expects a list.  
- The removal of the class Config with `validate_by_name` is redundant with `model_config`, which is the new pattern in Pydantic v2; this is good.
- Replacing string `example` with list `examples` improves documentation and validation clarity, aligning with Pydantic best practices.
- The use of `Dict` without specifying key types (e.g., `Dict[str, Any]`) is acceptable but explicit typing could enhance clarity.

**Improvements:**
- Ensure data validation remains intact with these changes.
- Explicitly typing dictionaries for improved correctness.

#### 3. **WhatsApp incoming message models**
- Transitioning to `ConfigDict` (Pydantic v2) is correct and future-proof.
- No significant issues, but consistency in the usage pattern should be ensured across models.

#### 4. **Test files (`test_vector_store_azure_vector_search.py`, `test_vector_store_chroma.py`, etc.)**
- Adding environment variables in tests is necessary, but doing so directly can be risky in shared environments. Consider using mock or fixture setup.
- Hardcoded API keys (`"sk-xxxx"`) is acceptable for tests but always ensure these are not committed or exposed publicly.

#### 5. **Application (`run.py`) Changes**
- Slight refactor from `mcp = FastMCP(stateless_http=True)` to `mcp = FastMCP()` is fine if the default is acceptable.
- The addition of `ws="websockets-sansio"` is good for local or testing environments.
- Suppressing verbose logging for Azure monitoring is a pragmatic improvement.

#### 6. **`generate.py` logging improvements**
- Wrapping print statements with a message-cleaning function (`clean_message_for_console`) is an effective pattern to avoid cluttering logs with binary or excessively verbose data.
- No issues here besides ensuring the function is efficient.

#### 7. **Utility Function (`utils.py`)**
- The cleaning logic is well designed, deep copying the message to avoid in-place modification.
- The condition to remove `additional_info["data"]` if it's binary is good but should have error handling if `additional_info` is not a dict or doesn't contain `"data"`.

**Potential refinement:**
```python
if message.message_context and isinstance(message.message_context.additional_info, dict):
    message.message_context.additional_info.pop("data", None)
```
To avoid potential KeyError.

#### 8. **Overall code structure and quality**
- Use of Pydantic v2 config and models is up-to-date.
- Code readability has improved, but some redundant comments or misplaced docstrings could be cleaned.
- Notable improvement: consistent use of `examples` in data models adds clarity for API consumers.

---

### Strengths
- Clear focus on type safety, especially in Pydantic models.
- Well-organized test setup modifications.
- Pragmatic logging suppression.
- Thoughtful utility function for message sanitization, preventing sensitive or binary data from leaking into logs.

### Areas for Improvement
- **Dependency Management:** Verify dependency caching especially in CI/CD workflows.
- **Error Handling:** The `clean_message_for_console()` should be more defensive against unexpected data types.
- **Testing Robustness:** Environment variable setting in tests can be risky; prefer mocking.
- **Documentation:** Some models could benefit from more detailed docstrings explaining the purpose of fields.
- **Reduce Duplication:** Patterns like setting `model_config` could be migrated into a base class to avoid repetition across models.

---

### Final Score: **75/100**

**Rationale:**  
The updates improve data typing and logging, reflecting good practices; however, some minor oversight in error handling, dependency management and testing environment safety reduce the score. With targeted improvements, this can reach a more robust level.

---

### Summary
This PR exhibits solid refactoring and modernization efforts, especially with Pydantic v2 features. It enhances code clarity and prepares the project for future updates. Still, attention to error handling, dependency reliability, and testing practices is necessary to elevate the overall quality.

---

**Web articles for reference:**
- [Pydantic V2 Release Notes](https://pydantic.dev/latest/changes/#model-config)
- [Effective Logging Control in Python](https://realpython.com/python-logging/)
- [Best Practices for CI/CD Environment Variable Management](https://docs.github.com/en/actions/security-guides/encrypted-secrets)
